### PR TITLE
Two improvements for parseq-zk-client

### PR DESCRIPTION
**Pull request by angxu**
This patch contains two improvements for parseq-zk-client.

1. Populate the path information from KeeperException thrown by #multi(Iterable)

2.  Add MultiLocks to zookeeper recipes

### CODE REVIEW

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java [hunk 1]
```diff
@@ -302,8 +304,34 @@ protected Promise run(Context context) throws Throwable {
    * @param executor {@code Executor} that will be used to run the operations.
    * @return task to execute multiple operations.
    */
-  public Task<List<OpResult>> multi(Iterable<Op> ops, Executor executor) {
-    return Task.blocking(() -> _zkClient.multi(ops), executor);
+  public Task<List<OpResult>> multi(List<Op> ops, Executor executor) {
```
Comment (zhenkai) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: This is technically backward incompatible. Not sure what's parseq's policy on this for contrib project

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java [hunk 2]
```diff
@@ -302,8 +304,34 @@ protected Promise run(Context context) throws Throwable {
...
+    return Task.blocking(() -> _zkClient.multi(ops), executor)
+        .recoverWith(e -> {
+          if (e instanceof KeeperException) {
+            e = getCause((KeeperException) e, ops);
```
Comment (zhenkai) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: do not reuse "e" here...

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java (hunk 1 above)
Comment (angxu) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: True, but it's not released yet anyway ;-p

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java [hunk 3]
```diff
@@ -302,8 +304,34 @@ protected Promise run(Context context) throws Throwable {
...
+  private KeeperException getCause(KeeperException ex, List<Op> ops) {
+    List<OpResult> results = ex.getResults();
+    for (int i = 0; i < results.size(); ++i) {
+      if (results.get(i) instanceof OpResult.ErrorResult &&
+          ((OpResult.ErrorResult) results.get(i)).getErr() != KeeperException.Code.OK.intValue()) {
+        OpResult.ErrorResult error = (OpResult.ErrorResult) results.get(i);
+        return KeeperException.create(KeeperException.Code.get(error.getErr()), ops.get(i).getPath());
+      }
+    }
+    throw new IllegalArgumentException("None of the Ops is failed. Ops results: " + ex.getResults());
```
Comment (zhenkai) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: how about just return ex?

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java [hunk 4]
```diff
@@ -340,12 +368,20 @@ protected Promise run(Context context) throws Throwable {
   public Task<String> ensurePathExists(String path) {
     return exists(path).flatMap(stat -> {
       if (!stat.isPresent()) {
+        Task<String> createAndIgnoreNodeExistsException =
```
Comment (zhenkai) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: createIfAbsent?

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/recipes/MultiLocks.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package com.linkedin.parseq.zk.recipes;
+
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.zk.client.ZKClient;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static java.util.stream.Collectors.*;
```
Comment (zhenkai) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/recipes/MultiLocks.java]: expand

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java (hunk 3 above)
Comment (angxu) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: In theory, we should never reach this line. So I'd rather throw exception here to (hopefully) raise more attention.

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java (hunk 1 above)
Comment (angxu) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/client/ZKClient.java]: After thinking a bit more, I decided to revert this change because it would erase the partial results of finished operations.

File: contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/recipes/Synchronizable.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package com.linkedin.parseq.zk.recipes;
+
+import com.linkedin.parseq.Task;
+
+
+/**
+ * @author Ang Xu
```
Comment (jodzga) [inline contrib/parseq-zk-client/src/main/java/com/linkedin/parseq/zk/recipes/Synchronizable.java]: Add javadoc.

**Comment by zhenkai** [comment]
LGTM. Merged.
